File: go/roles_drift_test.go
package realmid

import (
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"
)

// Cross-repo drift test for the two hand-maintained lists in roles_authority.go.
//
// THE ISSUER WINS. `ConfersAuthority` needs no list at all — it reads the
// ACTION off the permission string, so it cannot drift. These two CAN:
//
//	humanOnlyPermissions  ← issuer internal/realmrole/assignable.go
//	systemUnassignable    ← issuer internal/realmrole/store.go (NonAssignableRoles)
//
// This test parses the issuer's own source when a checkout is reachable and
// fails on any difference. That is the case in the RealmID workspace, where
// `sdk/` and `issuer/` are siblings, and in any CI that checks both out.
//
// WHEN THE ISSUER IS NOT REACHABLE the check cannot run, and it says so in
// large letters rather than reporting a pass it did not earn. `Realm-ID/sdk`'s
// own CI clones only this repo, so that is the standalone-CI case; wiring the
// issuer into it (or moving this check to the umbrella repo's cross-repo CI) is
// filed in sdk/TODO.md. A skipped gate is not a verdict — read the log line.
const driftCheckUnreachable = "" +
	"DRIFT CHECK DID NOT RUN: no issuer checkout found. " +
	"The lists in roles_authority.go were NOT compared against " +
	"issuer/internal/realmrole/. This is not a pass. See sdk/TODO.md."

// findIssuerRealmrole walks up from the working directory looking for a
// sibling issuer checkout. Returns "" when there is none.
func findIssuerRealmrole() string {
	dir, err := os.Getwd()
	if err != nil {
		return ""
	}
	for i := 0; i < 8; i++ {
		cand := filepath.Join(dir, "issuer", "internal", "realmrole")
		if st, err := os.Stat(filepath.Join(cand, "permissions.go")); err == nil && !st.IsDir() {
			return cand
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return ""
}

func parseIssuerFile(t *testing.T, path string) *ast.File {
	t.Helper()
	f, err := parser.ParseFile(token.NewFileSet(), path, nil, parser.ParseComments)
	if err != nil {
		// NOT a skip: the file is there and we could not read it. A gate that
		// treats input it cannot parse as input that passed is worse than none.
		t.Fatalf("parse issuer source %s: %v", path, err)
	}
	return f
}

// issuerPermConsts collects `PermX = "resource:action"` from permissions.go.
func issuerPermConsts(t *testing.T, f *ast.File) map[string]string {
	t.Helper()
	out := map[string]string{}
	for _, d := range f.Decls {
		gd, ok := d.(*ast.GenDecl)
		if !ok || gd.Tok != token.CONST {
			continue
		}
		for _, s := range gd.Specs {
			vs, ok := s.(*ast.ValueSpec)
			if !ok || len(vs.Names) != len(vs.Values) {
				continue
			}
			for i, n := range vs.Names {
				lit, ok := vs.Values[i].(*ast.BasicLit)
				if !ok || lit.Kind != token.STRING {
					continue
				}
				v, err := strconv.Unquote(lit.Value)
				if err != nil {
					continue
				}
				out[n.Name] = v
			}
		}
	}
	if len(out) == 0 {
		t.Fatalf("no Perm* constants found in the issuer's permissions.go — the parse silently matched nothing")
	}
	return out
}

// issuerMapKeys returns the composite-literal keys of a top-level `var name = map[...]...{...}`.
// Identifier keys come back as the identifier; string-literal keys as the string.
func issuerMapKeys(t *testing.T, f *ast.File, name string) []string {
	t.Helper()
	var keys []string
	found := false
	for _, d := range f.Decls {
		gd, ok := d.(*ast.GenDecl)
		if !ok || gd.Tok != token.VAR {
			continue
		}
		for _, s := range gd.Specs {
			vs, ok := s.(*ast.ValueSpec)
			if !ok {
				continue
			}
			for i, n := range vs.Names {
				if n.Name != name || i >= len(vs.Values) {
					continue
				}
				cl, ok := vs.Values[i].(*ast.CompositeLit)
				if !ok {
					t.Fatalf("issuer %s is not a composite literal any more — re-read the source", name)
				}
				found = true
				for _, el := range cl.Elts {
					kv, ok := el.(*ast.KeyValueExpr)
					if !ok {
						t.Fatalf("issuer %s has a non key/value element — re-read the source", name)
					}
					switch k := kv.Key.(type) {
					case *ast.Ident:
						keys = append(keys, k.Name)
					case *ast.BasicLit:
						v, err := strconv.Unquote(k.Value)
						if err != nil {
							t.Fatalf("issuer %s: unquote key %s: %v", name, k.Value, err)
						}
						keys = append(keys, v)
					default:
						t.Fatalf("issuer %s has an unrecognised key expression — re-read the source", name)
					}
				}
			}
		}
	}
	if !found {
		t.Fatalf("issuer var %s not found — it was renamed or moved; this SDK copy is now unanchored", name)
	}
	if len(keys) == 0 {
		t.Fatalf("issuer %s parsed to ZERO keys — refusing to compare against nothing", name)
	}
	return keys
}

func setEqual(t *testing.T, what string, got map[string]struct{}, want []string) {
	t.Helper()
	wantSet := map[string]struct{}{}
	for _, w := range want {
		wantSet[w] = struct{}{}
	}
	for w := range wantSet {
		if _, ok := got[w]; !ok {
			t.Errorf("%s DRIFT: the issuer has %q and the SDK does not — the issuer wins, add it", what, w)
		}
	}
	for g := range got {
		if _, ok := wantSet[g]; !ok {
			t.Errorf("%s DRIFT: the SDK has %q and the issuer does not — the issuer wins, remove it", what, g)
		}
	}
}

func TestRolePredicatesMatchTheIssuer(t *testing.T) {
	dir := findIssuerRealmrole()
	if dir == "" {
		t.Log(driftCheckUnreachable)
		return
	}
	t.Logf("drift check running against %s", dir)

	permsFile := parseIssuerFile(t, filepath.Join(dir, "permissions.go"))
	assignFile := parseIssuerFile(t, filepath.Join(dir, "assignable.go"))
	storeFile := parseIssuerFile(t, filepath.Join(dir, "store.go"))

	consts := issuerPermConsts(t, permsFile)

	// 1. humanOnlyPermissions ← realmrole.HumanOnlyPermissions (identifier keys).
	var wantHumanOnly []string
	for _, ident := range issuerMapKeys(t, assignFile, "HumanOnlyPermissions") {
		v, ok := consts[ident]
		if !ok {
			t.Fatalf("issuer HumanOnlyPermissions names %s, which is not a Perm* string constant", ident)
		}
		wantHumanOnly = append(wantHumanOnly, v)
	}
	setEqual(t, "humanOnlyPermissions", humanOnlyPermissions, wantHumanOnly)

	// 2. systemUnassignable ← realmrole.NonAssignableRoles (string-literal keys).
	setEqual(t, "systemUnassignable", systemUnassignable, issuerMapKeys(t, storeFile, "NonAssignableRoles"))
}

// TestConfersAuthorityRuleMatchesTheIssuer pins the RULE, not a list: the
// issuer derives "mutating" from `p.Action != "read"` over the ADR-074 catalog.
// Our copy derives it from the permission STRING's action segment, which is the
// same rule without needing the catalog. If the issuer ever stops deriving it
// that way, this SDK predicate stops being equivalent and must be revisited.
func TestConfersAuthorityRuleMatchesTheIssuer(t *testing.T) {
	dir := findIssuerRealmrole()
	if dir == "" {
		t.Log(driftCheckUnreachable)
		return
	}
	src, err := os.ReadFile(filepath.Join(dir, "permissions.go"))
	if err != nil {
		t.Fatalf("read issuer permissions.go: %v", err)
	}
	if !strings.Contains(string(src), `p.Action != "read"`) {
		t.Errorf(`issuer no longer derives the mutating set from p.Action != "read" — ` +
			`ConfersAuthority in roles_authority.go is no longer equivalent to the issuer's`)
	}
	if !strings.Contains(string(src), "func ConfersAuthority(perms []string) bool") {
		t.Errorf("issuer realmrole.ConfersAuthority changed shape — re-derive the SDK copy")
	}
}

// TestNoPerRoleMFAFloorInTheIssuer is the ADR-101 interlock. The console's
// predicate carried a required_mfa_methods floor; the SDK deliberately does
// not, because ADR-101 retired the field. If the issuer ever reinstates a
// role-level MFA gate on the assignability path, this SDK predicate is
// under-filtering and pickers will start offering choices that 400.
func TestNoPerRoleMFAFloorInTheIssuer(t *testing.T) {
	dir := findIssuerRealmrole()
	if dir == "" {
		t.Log(driftCheckUnreachable)
		return
	}
	gate := filepath.Join(dir, "..", "httpapi", "role_assignable.go")
	src, err := os.ReadFile(gate)
	if err != nil {
		t.Fatalf("read issuer role_assignable.go: %v", err)
	}
	body := string(src)
	i := strings.Index(body, "func requireRoleAssignableToKind(")
	if i < 0 {
		t.Fatalf("issuer requireRoleAssignableToKind not found — the enforcement point moved")
	}
	if strings.Contains(body[i:], "FirstUnsatisfiableMFAMethod(") {
		t.Errorf("the issuer's assignability gate applies a per-role MFA floor again; " +
			"IsRoleAssignableTo in roles_authority.go does not and is now under-filtering")
	}
}

File: ts/src/roles-drift.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { readFileSync, existsSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import {
  HUMAN_ONLY_PERMISSIONS,
  NON_ASSIGNABLE_ROLES,
  confersAuthority,
  isRoleAssignableTo,
  isRoleSeatable,
  PRINCIPAL_KINDS,
} from "./roles.js";
import { SSO_DOMAIN_METHODS, SSO_DOMAIN_STATUSES } from "./sso-domains.js";

/**
 * DRIFT GATE for the two rules `roles.ts` copies out of the issuer.
 *
 * THE ISSUER WINS. `issuer/internal/realmrole/permissions.go` (the ADR-074
 * catalog + `IsMutatingPermission`/`ConfersAuthority`) and
 * `issuer/internal/realmrole/assignable.go` (`HumanOnlyPermissions`,
 * `AssignableToKind`) and `issuer/internal/realmrole/store.go`
 * (`NonAssignableRoles`), and `issuer/internal/tenantdomain/tenantdomain.go`
 * (`IsValidMethod` / `IsValidStatus`) are authoritative.
 *
 * EVERY set this SDK mirrors is compared here, by SET EQUALITY rather than
 * membership, so an extra entry fails as loudly as a missing one. That rule was
 * added after `NON_ASSIGNABLE_ROLES` shipped a member short while this gate was
 * green: the gate covered two of the mirrored sets and said nothing about the
 * third, which is indistinguishable from having no gate for that set at all. If this file goes red, the SDK is
 * wrong and the SDK is what changes.
 *
 * It binds in two independent ways, on purpose:
 *
 *  1. Against a PINNED FIXTURE below — a snapshot of the issuer's catalog,
 *     checked in as TEST DATA. This half runs everywhere, including a
 *     standalone `Realm-ID/sdk` CI checkout where no issuer source exists. It
 *     proves the SDK's colon-derived action rule classifies every real catalog
 *     entry exactly as the issuer's `Action != "read"` rule does.
 *  2. Against the LIVE issuer source, when a sibling checkout is reachable
 *     (the `Realm-ID/project` workspace layout, or `REALMID_ISSUER_DIR`). This
 *     is the half that catches the fixture going stale.
 *
 * Half 2 cannot bind in `Realm-ID/sdk`'s own CI, which checks out one repo —
 * so it says so out loud rather than passing quietly, and `REALMID_DRIFT_STRICT=1`
 * turns "issuer not reachable" into a failure for any runner that CAN reach it.
 * Wiring that checkout into `sdk`'s CI is filed in `sdk/TODO.md`.
 */

// ---- The pinned snapshot (regenerate from the issuer, never hand-edit) ----

interface CatalogEntry {
  key: string;
  resource: string;
  action: string;
}

/** `realmrole.Catalog`, as of issuer @ 2026-08-30. 31 entries (ADR-074). */
const ISSUER_CATALOG: CatalogEntry[] = [
  { key: "users:read", resource: "users", action: "read" },
  { key: "users:manage", resource: "users", action: "manage" },
  { key: "invitations:read", resource: "invitations", action: "read" },
  { key: "invitations:manage", resource: "invitations", action: "manage" },
  { key: "roles:read", resource: "roles", action: "read" },
  { key: "tenants:read", resource: "tenants", action: "read" },
  { key: "tenants:manage", resource: "tenants", action: "manage" },
  { key: "service_accounts:read", resource: "service_accounts", action: "read" },
  { key: "service_accounts:manage", resource: "service_accounts", action: "manage" },
  { key: "platform_api_keys:read", resource: "platform_api_keys", action: "read" },
  { key: "platform_api_keys:manage", resource: "platform_api_keys", action: "manage" },
  { key: "identity_providers:read", resource: "identity_providers", action: "read" },
  { key: "identity_providers:manage", resource: "identity_providers", action: "manage" },
  { key: "sources:read", resource: "sources", action: "read" },
  { key: "sources:manage", resource: "sources", action: "manage" },
  { key: "federation:read", resource: "federation", action: "read" },
  { key: "federation:manage", resource: "federation", action: "manage" },
  { key: "signing_keys:read", resource: "signing_keys", action: "read" },
  { key: "signing_keys:rotate", resource: "signing_keys", action: "rotate" },
  { key: "domains:read", resource: "domains", action: "read" },
  { key: "domains:manage", resource: "domains", action: "manage" },
  { key: "sessions:revoke", resource: "sessions", action: "revoke" },
  { key: "audit:read", resource: "audit", action: "read" },
  { key: "platform:config", resource: "platform", action: "config" },
  { key: "otp:read", resource: "otp", action: "read" },
  { key: "integrations:read", resource: "integrations", action: "read" },
  { key: "integrations:manage", resource: "integrations", action: "manage" },
  { key: "org_grants:read", resource: "org_grants", action: "read" },
  { key: "org_grants:manage", resource: "org_grants", action: "manage" },
  { key: "user_api_keys:read", resource: "user_api_keys", action: "read" },
  { key: "user_api_keys:manage", resource: "user_api_keys", action: "manage" },
];

/**
 * `realmrole.NonAssignableRoles` (store.go), sorted — the roles nobody is
 * granted through the invite/assignment surface.
 *
 * DELIBERATELY NOT `ProtectedRoles`, which the issuer keeps as a separate map
 * meaning "cannot be disabled or deleted". The two overlap but `member` is
 * protected AND the most assignable role there is, so reading one for the other
 * empties every picker.
 */
const ISSUER_NON_ASSIGNABLE = ["owner", "platform_api", "platform_mgmt_api"];

/** `realmrole.AssignableKinds` (ADR-071 `users.kind`), sorted. */
const ISSUER_ASSIGNABLE_KINDS = ["human", "service"];

/** `tenantdomain.IsValidMethod`'s five cases (ADR-094), sorted. */
const ISSUER_SSO_METHODS = [
  "dns_txt",
  "html_file",
  "meta_tag",
  "platform_approval",
  "self_asserted",
];

/** `tenantdomain.IsValidStatus`'s seven cases (ADR-094), sorted. */
const ISSUER_SSO_STATUSES = [
  "active",
  "claimed",
  "failed",
  "pending",
  "rejected",
  "revoked",
  "suspended",
];

/** `realmrole.HumanOnlyPermissions` (ADR-081 §2.3), sorted. */
const ISSUER_HUMAN_ONLY = [
  "domains:manage",
  "federation:manage",
  "platform:config",
  "signing_keys:rotate",
];

// ---- Half 1: the SDK agrees with the pinned catalog, everywhere ----

test("drift: the SDK's colon-derived action equals the issuer's catalog Action", () => {
  // This is the invariant that makes `confersAuthority` correct WITHOUT the
  // SDK shipping a copy of the catalog: for every real entry, the substring
  // after the colon IS the issuer's `Action` field. Break it in the issuer and
  // this fixture stops matching (half 2), and this assertion pins why it
  // matters.
  for (const p of ISSUER_CATALOG) {
    assert.equal(p.key, `${p.resource}:${p.action}`, `catalog key ${p.key} is not resource:action`);
    const expected = p.action !== "read";
    assert.equal(
      confersAuthority({ permissions: [p.key] }),
      expected,
      `${p.key}: issuer says confersAuthority=${expected}`,
    );
    assert.equal(
      confersAuthority({ permissions: [p.key] }, { catalog: ISSUER_CATALOG.map((c) => ({ ...c, label: "" })) }),
      expected,
      `${p.key}: catalog-driven classification must match`,
    );
  }
});

test("drift: the catalog has the 31 entries ADR-074 pins", () => {
  // The issuer pins the count AND the key set (TestCatalogSize + catalogKeys).
  // A swap that keeps the count is caught by half 2's set comparison.
  assert.equal(ISSUER_CATALOG.length, 31);
});

test("drift: HUMAN_ONLY_PERMISSIONS matches the pinned issuer set", () => {
  assert.deepEqual([...HUMAN_ONLY_PERMISSIONS].sort(), ISSUER_HUMAN_ONLY);
  // and every one of them is a real catalog key
  const keys = new Set(ISSUER_CATALOG.map((p) => p.key));
  for (const p of ISSUER_HUMAN_ONLY) assert.ok(keys.has(p), `${p} is not in the catalog`);
});

test("drift: NON_ASSIGNABLE_ROLES matches the pinned issuer set", () => {
  assert.deepEqual([...NON_ASSIGNABLE_ROLES].sort(), ISSUER_NON_ASSIGNABLE);
  // and the picker predicate actually applies it, for both kinds
  for (const name of ISSUER_NON_ASSIGNABLE) {
    assert.equal(isRoleSeatable({ name }, "human"), false, name);
    assert.equal(isRoleSeatable({ name }, "service"), false, name);
  }
});

test("drift: PRINCIPAL_KINDS matches the pinned issuer set", () => {
  assert.deepEqual([...PRINCIPAL_KINDS].sort(), ISSUER_ASSIGNABLE_KINDS);
});

test("drift: the SSO domain vocabularies match the pinned issuer sets", () => {
  assert.deepEqual([...SSO_DOMAIN_METHODS].sort(), ISSUER_SSO_METHODS);
  assert.deepEqual([...SSO_DOMAIN_STATUSES].sort(), ISSUER_SSO_STATUSES);
});

test("drift: every human-only permission is one a service principal is refused", () => {
  for (const p of ISSUER_HUMAN_ONLY) {
    assert.equal(
      isRoleAssignableTo({ name: "custom", permissions: [p], assignable_to: ["human", "service"] }, "service"),
      false,
    );
  }
});

// ---- Half 2: the pinned fixture still matches the live issuer source ----

function findIssuerDir(): string | null {
  const explicit = process.env["REALMID_ISSUER_DIR"];
  if (explicit) return existsSync(resolve(explicit, "internal")) ? resolve(explicit) : null;
  // sdk/ts/src -> sdk/ts -> sdk -> <workspace>
  const here = dirname(fileURLToPath(import.meta.url));
  const p = resolve(here, "../../../issuer");
  return existsSync(resolve(p, "internal/realmrole")) ? p : null;
}

/**
 * Keys of a Go map literal, addressed by its VARIABLE NAME.
 *
 * Anchored on the name and never on `map[string]bool{` alone: `store.go`
 * declares `ProtectedRoles` beside `NonAssignableRoles` with an identical type
 * and a different meaning, and it contains `member`. A loose match would swap
 * "cannot be deleted" for "cannot be held" and silently empty every picker.
 */
function issuerMapKeys(src: string, varName: string, where: string): string[] {
  const block = src.match(
    new RegExp(`var ${varName} = map\\[string\\](?:bool|struct\\{\\})\\{([\\s\\S]*?)\\n\\}`),
  );
  assert.ok(block, `could not locate \`var ${varName}\` in ${where}`);
  const keys = [...(block[1] as string).matchAll(/"([a-z0-9_:]+)":/g)].map((m) => m[1] as string);
  assert.ok(keys.length > 0, `parsed zero keys from ${varName} — the regex has stopped matching`);
  return keys.sort();
}

/** The string cases of a Go `switch` inside the named func — a closed vocabulary. */
function issuerSwitchCases(src: string, funcName: string, where: string): string[] {
  const fn = src.match(new RegExp(`func ${funcName}\\([^)]*\\) bool \\{([\\s\\S]*?)\\n\\}`));
  assert.ok(fn, `could not locate \`func ${funcName}\` in ${where}`);
  const caseLine = (fn[1] as string).match(/\n\tcase ([\s\S]*?):/);
  assert.ok(caseLine, `\`${funcName}\` has no case clause`);
  // The cases name Go consts (MethodDNSTXT, StatusClaimed); resolve each to its
  // declared string literal rather than guessing from the identifier.
  const idents = (caseLine[1] as string).split(",").map((x) => x.trim()).filter(Boolean);
  assert.ok(idents.length > 0, `parsed zero cases from ${funcName} — the regex has stopped matching`);
  const consts = new Map<string, string>();
  for (const m of src.matchAll(/^\s*([A-Z][A-Za-z0-9_]*)\s+(?:Method|Status)\s*=\s*"([^"]+)"/gm)) {
    consts.set(m[1] as string, m[2] as string);
  }
  return idents
    .map((id) => {
      const v = consts.get(id);
      assert.ok(v, `${funcName} names ${id}, which is not a declared const in ${where}`);
      return v;
    })
    .sort();
}

function parseIssuer(root: string): {
  catalog: CatalogEntry[];
  humanOnly: string[];
  nonAssignable: string[];
  assignableKinds: string[];
  ssoMethods: string[];
  ssoStatuses: string[];
} {
  const dir = resolve(root, "internal/realmrole");
  const perms = readFileSync(resolve(dir, "permissions.go"), "utf8");
  const assignable = readFileSync(resolve(dir, "assignable.go"), "utf8");
  const store = readFileSync(resolve(dir, "store.go"), "utf8");
  const td = readFileSync(resolve(root, "internal/tenantdomain/tenantdomain.go"), "utf8");

  const consts = new Map<string, string>();
  for (const m of perms.matchAll(/^\s*(Perm[A-Za-z0-9_]+)\s*=\s*"([^"]+)"/gm)) {
    consts.set(m[1] as string, m[2] as string);
  }
  assert.ok(consts.size >= 25, `parsed only ${consts.size} Perm consts — the regex has stopped matching`);

  const block = perms.match(/var Catalog = \[\]Permission\{([\s\S]*?)\n\}/);
  assert.ok(block, "could not locate `var Catalog` in permissions.go");
  const catalog: CatalogEntry[] = [];
  for (const m of (block[1] as string).matchAll(/\{(Perm[A-Za-z0-9_]+),\s*"([^"]*)",\s*"([^"]*)",/g)) {
    const key = consts.get(m[1] as string);
    assert.ok(key, `catalog names ${m[1]}, which is not a declared Perm const`);
    catalog.push({ key, resource: m[2] as string, action: m[3] as string });
  }
  assert.ok(catalog.length >= 25, `parsed only ${catalog.length} catalog entries — the regex has stopped matching`);

  const hoBlock = assignable.match(/var HumanOnlyPermissions = map\[string\]struct\{\}\{([\s\S]*?)\n\}/);
  assert.ok(hoBlock, "could not locate `var HumanOnlyPermissions` in assignable.go");
  const humanOnly: string[] = [];
  for (const m of (hoBlock[1] as string).matchAll(/^\s*(Perm[A-Za-z0-9_]+):/gm)) {
    const key = consts.get(m[1] as string);
    assert.ok(key, `HumanOnlyPermissions names ${m[1]}, which is not a declared Perm const`);
    humanOnly.push(key);
  }
  assert.ok(humanOnly.length > 0, "parsed zero human-only permissions — the regex has stopped matching");

  const nonAssignable = issuerMapKeys(store, "NonAssignableRoles", "store.go");
  assert.ok(
    !nonAssignable.includes("member"),
    "parsed `member` as non-assignable — the reader matched ProtectedRoles, not NonAssignableRoles",
  );

  const kindsBlock = assignable.match(/var AssignableKinds = \[\]string\{([^}]*)\}/);
  assert.ok(kindsBlock, "could not locate `var AssignableKinds` in assignable.go");
  const kindConsts = new Map<string, string>();
  for (const m of assignable.matchAll(/^\s*(Assignable[A-Za-z]+)\s*=\s*"([^"]+)"/gm)) {
    kindConsts.set(m[1] as string, m[2] as string);
  }
  const assignableKinds = (kindsBlock[1] as string)
    .split(",")
    .map((x) => x.trim())
    .filter(Boolean)
    .map((id) => {
      const v = kindConsts.get(id);
      assert.ok(v, `AssignableKinds names ${id}, which is not a declared const`);
      return v;
    })
    .sort();
  assert.ok(assignableKinds.length > 0, "parsed zero assignable kinds — the regex has stopped matching");

  return {
    catalog,
    humanOnly: humanOnly.sort(),
    nonAssignable,
    assignableKinds,
    ssoMethods: issuerSwitchCases(td, "IsValidMethod", "tenantdomain.go"),
    ssoStatuses: issuerSwitchCases(td, "IsValidStatus", "tenantdomain.go"),
  };
}

test("drift: the pinned fixture equals the LIVE issuer source", (t) => {
  const dir = findIssuerDir();
  if (!dir) {
    // Not a pass on the merits and not silent about it. Half 1 above still
    // bound; only the fixture-staleness check did not.
    const msg =
      "issuer checkout not reachable — the fixture-staleness half did NOT run. " +
      "Set REALMID_ISSUER_DIR=<path to Realm-ID/issuer> to bind it.";
    if (process.env["REALMID_DRIFT_STRICT"]) assert.fail(msg);
    t.diagnostic(msg);
    return;
  }
  const live = parseIssuer(dir);
  assert.deepEqual(live.catalog, ISSUER_CATALOG, "the ADR-074 catalog moved; regenerate ISSUER_CATALOG");
  assert.deepEqual(live.humanOnly, ISSUER_HUMAN_ONLY, "HumanOnlyPermissions moved; update roles.ts AND this fixture");
  assert.deepEqual(
    live.nonAssignable,
    ISSUER_NON_ASSIGNABLE,
    "NonAssignableRoles moved; update NON_ASSIGNABLE_ROLES in roles.ts AND this fixture",
  );
  assert.deepEqual(
    live.assignableKinds,
    ISSUER_ASSIGNABLE_KINDS,
    "AssignableKinds moved; update PRINCIPAL_KINDS in roles.ts AND this fixture",
  );
  assert.deepEqual(
    live.ssoMethods,
    ISSUER_SSO_METHODS,
    "tenantdomain methods moved; update SSO_DOMAIN_METHODS AND this fixture",
  );
  assert.deepEqual(
    live.ssoStatuses,
    ISSUER_SSO_STATUSES,
    "tenantdomain statuses moved; update SSO_DOMAIN_STATUSES AND this fixture",
  );
});
